File: src/main/java/cc/sukazyo/cono/morny/bot/query/ShareToolTwitter.java
package cc.sukazyo.cono.morny.bot.query;

import cc.sukazyo.cono.morny.bot.api.InlineQueryUnit;
import com.pengrad.telegrambot.model.Update;
import com.pengrad.telegrambot.model.request.InlineQueryResultArticle;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import static cc.sukazyo.cono.morny.util.tgapi.formatting.NamedUtils.inlineIds;

public class ShareToolTwitter implements ITelegramQuery {
	
	public static final String TITLE_VX = "[tweet] Share as VxTwitter";
	public static final String TITLE_VX_COMBINED = "[tweet] Share as VxTwitter(combination)";
	public static final String ID_PREFIX_VX = "[morny/share/twitter/vxtwi]";
	public static final String ID_PREFIX_VX_COMBINED = "[morny/share/twitter/vxtwi_combine]";
	
	public static final Pattern REGEX_TWEET_LINK = Pattern.compile(
			"^(?:https?://)?((?:(?:c\\.)?vx|fx|www\\.)?twitter\\.com)/((\\w+)/status/(\\d+)(?:/photo/(\\d+))?)/?(\\?[\\w&=-]+)?$");
	
	@Nullable
	@Override
	public List<InlineQueryUnit<?>> query (@Nonnull Update event) {
		if (event.inlineQuery().query() == null) return null;
		final Matcher regex = REGEX_TWEET_LINK.matcher(event.inlineQuery().query());
		if (regex.matches()) {
			
			List<InlineQueryUnit<?>> result = new ArrayList<>();
			
			result.add(new InlineQueryUnit<>(new InlineQueryResultArticle(
					inlineIds(ID_PREFIX_VX+event.inlineQuery().query()), TITLE_VX,
					String.format("https://vxtwitter.com/%s", regex.group(2))
			)));
			result.add(new InlineQueryUnit<>(new InlineQueryResultArticle(
					inlineIds(ID_PREFIX_VX_COMBINED+event.inlineQuery().query()), TITLE_VX_COMBINED,
					String.format("https://c.vxtwitter.com/%s", regex.group(2))
			)));
			
			return result;
			
		}
		return null;
	}
	
}

File: src/main/scala/cc/sukazyo/cono/morny/bot/query/ShareToolTwitter.scala
package cc.sukazyo.cono.morny.bot.query

import cc.sukazyo.cono.morny.extra.twitter
import cc.sukazyo.cono.morny.extra.twitter.TweetUrlInformation
import com.pengrad.telegrambot.model.Update
import com.pengrad.telegrambot.model.request.InlineQueryResultArticle

import scala.language.postfixOps

class ShareToolTwitter extends ITelegramQuery {
	
	private val TITLE_VX = "[Twitter/X][VxTwitter]"
	private val ID_PREFIX_VX = "[morny/share/twitter/vx_url]"
	private val TITLE_FX = "[Twitter/X][Fix-Tweet]"
	private val ID_PREFIX_FX = "[morny/share/twitter/fx_url]"
	
	override def query (event: Update): List[InlineQueryUnit[_]] | Null = {
		
		if (event.inlineQuery.query == null) return null
		
		def getQueryTweetId (prefix: String, tweet: TweetUrlInformation): String =
			prefix + tweet.hashCode
		def getTweetName (title_prefix: String, tweet: TweetUrlInformation): String =
			s"$title_prefix ${tweet.screenName}/${tweet.statusId}"
		
		twitter.guessTweetUrl(event.inlineQuery.query).flatMap(tweet =>
			List(
				InlineQueryUnit(InlineQueryResultArticle(
					getQueryTweetId(ID_PREFIX_FX, tweet),
					getTweetName(TITLE_FX, tweet),
					s"https://fxtwitter.com/${tweet.statusPath}"
				)),
				InlineQueryUnit(InlineQueryResultArticle(
					getQueryTweetId(ID_PREFIX_VX, tweet),
					getTweetName(TITLE_VX, tweet),
					s"https://vxtwitter.com/${tweet.statusPath}"
				))
			)
		)
	}
	
}
